eagerly force arguments to builtin functions

o- opened the pull request:
see individual commit msgs.


We missed this in the past, so we would actually pass unevaluated promises. (in short this means that speculating on builtins was mostly broken :) )

but, for that reason I had to move the guard before the evaluation of the promises, because of course if we force an argument, then find out that we are actually not calling a builtin after all, then the forcing would have been wrong.

o- committed:
bugfix: actually use the replaced value

o- committed:
fix call argument types

call instruction really needs promises, not something else

o- committed:
eager promise values can always be inlined, since they are already
forced

o- committed:
call builtin always takes value

builtins are eager...

o- committed:
builtins are eager

if we speculate on builtins, we should eagerly evaluate all args.
This was missing before, so we would actually pass unevaluated
promises.

To be able to do this, we have to put the guard earlier, ie. before
the arguments. because if it turns out not to be a builtin, forcing
early would be wrong.

o- committed:
c is also semi-safe builtin

charig commented on rir/src/compiler/opt/elide_env.cpp:
```diff
@@ -45,8 +45,10 @@ void ElideEnv::apply(RirCompiler&, Closure* function, LogStream&) const {
+                        auto safe =
+                            new CallSafeBuiltin(b->blt, args, b->srcIdx);
+                        b->replaceUsesWith(safe);
+                        bb->replace(ip, safe);
```
Just wondering...Is there any case in which we replace and do not want to replace uses?

o- replied:
hmm, probably zero...

o- replied:
ok, I tried to change it. but there are some places (e.g. in lower expect) where the concerns (replacing uses vs. replacing instructions) are separated between different parts of the code. so it's not a simple change and thus I'm not gonna add it here.

charig reviewed (approved):
Changes seems ok modulo build is failing. Most were already approved in another PR.
The changes in the processing of calls in `rir_2_pir` will clash with #292.

o- commented:
>  The changes in the processing of calls in rir_2_pir will clash with #292.

yeah, that's because I had to reorder the guard. but it should subsume your refactoring, as this change also pulls the guard creation out and avoids the duplicated code

o- committed:
creating the promise should not be necessary in most cases

o- committed:
fix issue with checkpoints in promises

when we eagerly inline a promise we should avoid creating checkpoints.
The logic there is wrong, the checkpoint would have to re-create the
promise at runtime (which currently is not possible).
